### PR TITLE
EigenDA Disperser Store Module

### PR DESCRIPTION
**Pull request by gianbelinche**
## What ❔
This PR adds the Disperser store module (EigenDAClient), that communicates with the disperser grpc

It implements the `put_blob` and `get_blob` functions, that will later be called by the `/get` and `/put` endpoints.

In order to eventually run this, the config should be provided as followed on `etc/env/file_based/general.yaml`

```
da_client:
  eigen_da:
    disperser:
      api_node_url: http://127.0.0.1:4242 # TODO: This should be removed once eigenda proxy is no longer used
      disperser_rpc: <your_desired_disperser>
      eth_confirmation_depth: -1
      eigenda_eth_rpc: <your_desired_rpc>
      eigenda_svc_manager_addr: <your_svc_manager_addr>
      blob_size_limit: 2097152
      status_query_timeout: 1800
      status_query_interval: 5
      wait_for_finalization: false
```

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zk_supervisor fmt` and `zk_supervisor lint`.

### CODE REVIEW

File: core/node/eigenda_proxy/src/eigenda_proxy_client.rs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+use tokio::{sync::Mutex, time::interval};
+use tonic::transport::{Channel, ClientTlsConfig};
+use zksync_config::configs::da_client::eigen_da::DisperserConfig;
+
+use crate::{
+    blob_info::BlobInfo,
+    disperser::{self, disperser_client::DisperserClient, BlobStatusRequest, DisperseBlobRequest},
+};
+
+pub struct EigenDAProxyClient {
```
Comment (juanbono) [inline core/node/eigenda_proxy/src/eigenda_proxy_client.rs:17]: I don't think it should be named EigenDAProxyClient since there is no proxy here.

File: core/node/eigenda_proxy/Cargo.toml [hunk 2]
```diff
@@ -17,11 +17,14 @@ tracing.workspace = true
 rlp.workspace = true
 rand.workspace = true
 sha3.workspace = true
+zksync_config.workspace = true
 # we can't use the workspace version of prost because
 # the tonic dependency requires a hugher version.
 prost = "0.13.1"
-tonic = { version = "0.12.1", features = ["tls", "channel", "tls-roots"]}
+tonic = { version = "0.12.1", features = ["tls", "channel", "tls-native-roots"]}
```
Comment (juan518munoz) [inline core/node/eigenda_proxy/Cargo.toml:24]: What's the reason behind this change? 🤔

Comment (gianbelinche) [reply core/node/eigenda_proxy/Cargo.toml:24]: It is deprecated
```
tls-roots: Deprecated. An alias to tls-native-roots feature.
tls-native-roots: Adds system trust roots to [rustls](https://docs.rs/rustls)-based gRPC clients using the [rustls-native-certs](https://docs.rs/rustls-native-certs) crate. Not enabled by default.
```
From https://docs.rs/tonic/latest/tonic/index.html

File: core/node/eigenda_proxy/src/blob_info.rs [hunk 3]
```diff
@@ -35,6 +60,16 @@ impl Encodable for G1Commitment {
...
 
+impl TryFrom<DisperserG1Commitment> for G1Commitment {
+    type Error = ConversionError;
+    fn try_from(value: DisperserG1Commitment) -> Result<Self, Self::Error> {
+        Ok(Self {
+            x: value.x,
+            y: value.y,
+        })
+    }
+}
```
Comment (juan518munoz) [inline core/node/eigenda_proxy/src/blob_info.rs:71]: This could be replaced for a simpler `From` implementation:
```rs
impl From<G1Commitment> for DisperserG1Commitment {
    fn from(value: G1Commitment) -> Self {
        Self {
            x: value.x,
            y: value.y,
        }
    }
}
```

Casting from one type to the other in not error prone.

File: core/node/eigenda_proxy/src/blob_info.rs [hunk 4]
```diff
@@ -76,6 +111,18 @@ impl Encodable for BlobQuorumParam {
...
+    type Error = ConversionError;
+    fn try_from(value: DisperserBlobQuorumParam) -> Result<Self, Self::Error> {
+        Ok(Self {
+            quorum_number: value.quorum_number,
+            adversary_threshold_percentage: value.adversary_threshold_percentage,
+            confirmation_threshold_percentage: value.confirmation_threshold_percentage,
+            chunk_length: value.chunk_length,
+        })
+    }
+}
```
Comment (juan518munoz) [inline core/node/eigenda_proxy/src/blob_info.rs:124]: Same as above:
```rs
impl From<BlobQuorumParam> for DisperserBlobQuorumParam {
    fn from(value: BlobQuorumParam) -> Self {
        Self {
            quorum_number: value.quorum_number,
            adversary_threshold_percentage: value.adversary_threshold_percentage,
            confirmation_threshold_percentage: value.confirmation_threshold_percentage,
            chunk_length: value.chunk_length,
        }
    }
}
```

File: core/node/eigenda_proxy/src/blob_info.rs [hunk 5]
```diff
@@ -165,6 +232,18 @@ impl Encodable for BatchHeader {
...
+    type Error = ConversionError;
+    fn try_from(value: DisperserBatchHeader) -> Result<Self, Self::Error> {
+        Ok(Self {
+            batch_root: value.batch_root,
+            quorum_numbers: value.quorum_numbers,
+            quorum_signed_percentages: value.quorum_signed_percentages,
+            reference_block_number: value.reference_block_number,
+        })
+    }
+}
```
Comment (juan518munoz) [inline core/node/eigenda_proxy/src/blob_info.rs:245]: Ditto:
```rs
impl From<BatchHeader> for DisperserBatchHeader {
    fn from(value: BatchHeader) -> Self {
        Self {
            batch_root: value.batch_root,
            quorum_numbers: value.quorum_numbers,
            quorum_signed_percentages: value.quorum_signed_percentages,
            reference_block_number: value.reference_block_number,
        }
    }
}
```

File: core/node/eigenda_proxy/src/eigenda_client.rs [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+pub struct EigenDAClient {
+    disperser: Arc<Mutex<DisperserClient<Channel>>>,
+    config: DisperserConfig,
+}
+
+impl EigenDAClient {
+    pub async fn new(config: DisperserConfig) -> Result<Self, EigenDAError> {
+        match rustls::crypto::ring::default_provider().install_default() {
+            Ok(_) => {}
+            Err(_) => {} // This is not an actual error, we expect this function to return an Err(Arc<CryptoProvider>)
```
Comment (juan518munoz) [inline core/node/eigenda_proxy/src/eigenda_client.rs:26]: Comment is somewhat unclear, why aren't we making use of this `Arc<CryptoProvider>` maybe something like this is more suitable:
```rs        
        // This might fail if the default provider is already installed
        // So we ignore the result altogether
        let _ = rustls::crypto::ring::default_provider().install_default();
```

File: core/node/eigenda_proxy/src/eigenda_client.rs [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+            0 => disperser::BlobStatus::Unknown,
+            1 => disperser::BlobStatus::Processing,
+            2 => disperser::BlobStatus::Confirmed,
+            3 => disperser::BlobStatus::Failed,
+            4 => disperser::BlobStatus::Finalized,
+            5 => disperser::BlobStatus::InsufficientSignatures,
+            6 => disperser::BlobStatus::Dispersing,
+            _ => disperser::BlobStatus::Unknown,
+        }
+    }
```
Comment (juan518munoz) [inline core/node/eigenda_proxy/src/eigenda_client.rs:57]: `BlobStatus` is `#[repr(i32)]`, so this function can be replaced with a inherited `try_from`:
```rs
        match disperser::BlobStatus::try_from(result) {
            Ok(status) => status,
            Err(_) => disperser::BlobStatus::Unknown,
        }
```

**Review by juan518munoz — CHANGES_REQUESTED** [review]
PR looks good overall, but let's discuss the comments I left before approval.

**Review by juan518munoz — APPROVED** [review]
LGTM

**Review by juanbono — APPROVED** [review]
